### PR TITLE
(Feat): Interacting with projects programmatically

### PR DESCRIPTION
**Pull request by DanielHougaard**
## Description 📣

This PR introduces a new concept called ghost users. Due to the nature of Infisical's end-to-end encryption structure, it was previously not possible to interact with projects programmatically through the API and other consuming applications. By introducing ghost user's, we're slowly moving to a more industry-standard way of storing secrets securely, while at the same time improving the user experience for everyone.
To sum it up. The main purpose of the PR is to make the API more user-friendly, and to introduce a better approach as to how we securely store secrets. Existing projects will get an option to upgrade to the new ghost user logic. All new projects will be created with the ghost logic by default.

## Type ✨

- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation

---

- [x] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

File: frontend/src/views/SecretApprovalPage/components/SecretApprovalRequest/components/SecretApprovalRequestChanges.tsx
Comment (akhilmhdh) [inline frontend/src/views/SecretApprovalPage/components/SecretApprovalRequest/components/SecretApprovalRequestChanges.tsx:12]: use twmerge for this.

File: frontend/src/pages/org/[id]/overview/index.tsx
Comment (akhilmhdh) [inline frontend/src/pages/org/[id]/overview/index.tsx:523]: Didn't understand.

File: frontend/src/components/v2/UpgradeProjectAlert/UpgradeProjectAlert.tsx
Comment (akhilmhdh) [inline frontend/src/components/v2/UpgradeProjectAlert/UpgradeProjectAlert.tsx:54]: You can just interval feature in react-query. You don't need to manually do this. Based on the return api u can output it directly. You don't need to use another derived state

**Review by akhilmhdh — CHANGES_REQUESTED** [review]
One more question. As we are polling the system. Are we blocking user interactions. 


Because what will happen if user's does something while the indexing is happening behind the scene. Blind index didn't had this issue because those secrets were never accessible by end users. But this case is not

**Comment by DanielHougaard** [comment]
> One more question. As we are polling the system. Are we blocking user interactions.
> 
> Because what will happen if user's does something while the indexing is happening behind the scene. Blind index didn't had this issue because those secrets were never accessible by end users. But this case is not

This is an interesting case yea. I agree, we should to some degree block secret mutations while an upgrade is happening. I've just pushed a change that blocks secret write operations if an upgrade is active. You can check it out [here](https://github.com/Infisical/infisical/pull/1400/commits/1a799e31ca57b3fb1cbb56461eb81eb00eb842c6).

File: backend/src/server/routes/v2/index.ts
Comment (DanielHougaard) [reply backend/src/server/routes/v2/index.ts:25]: Last I spoke to Tony, I was told that we'd be doing a "larger" move at a later time, where we'll properly start renaming things to the `project` namespace. The reason for going with the `v2` router, is because the `v1` router would have conflicting endpoints if we use the `workspace` namespace. I think if we're planning to do a larger renaming later, we should keep this on the `v2` router for this.

File: backend/src/services/org/org-service.ts
Comment (DanielHougaard) [reply backend/src/services/org/org-service.ts:136]: > Use the alphaNum nanoId in lib. The above will generate symbols which would practically make the email invalid.

The only type of special characters that `nanoid` uses, is `_` and `-`, which is fine.

> This would ensure if there is any point of email verification being done its still a valid email. Another thing use @infisical because that would reduce scope of a domain attack.

Email verification won't be done, as it's not possible to log in as the ghost user at any point. However I've added `@infisical`.

File: backend/src/services/project-membership/project-membership-service.ts
Comment (DanielHougaard) [reply backend/src/services/project-membership/project-membership-service.ts:80]: The function was like this before my changes as well, and in fact this endpoint isn't actually used anywhere. We have a query for it in the frontend, but that query isn't actually used anywhere at all. However, I've updated the function now to address the mentioned issues.

File: backend/src/services/project/project-dal.ts
Comment (DanielHougaard) [reply backend/src/services/project/project-dal.ts:71]: Yeah, but I'd like an explicit function for this for two reasons.

1. A custom error message that reflects exactly which DAL operation failed
2. For readability and reusability. 

I would prefer to keep it this way.

File: backend/src/ee/services/secret-approval-request/secret-approval-request-secret-dal.ts
Comment (DanielHougaard) [reply backend/src/ee/services/secret-approval-request/secret-approval-request-secret-dal.ts:26]: We are actually only getting 100 versions per folder, the rest are discarded.

If I was going to use the `insert` method you mentioned, I'd need a full object instead of just a partial update object.

Do you think this will still be an issue, considering we're only updating 100 versions per folder?

File: backend/src/services/secret/secret-dal.ts
Comment (DanielHougaard) [reply backend/src/services/secret/secret-dal.ts:53]: Very good catch!

File: backend/src/services/secret/secret-version-dal.ts
Comment (DanielHougaard) [reply backend/src/services/secret/secret-version-dal.ts:49]: Very good catch!

File: frontend/src/pages/org/[id]/overview/index.tsx
Comment (DanielHougaard) [reply frontend/src/pages/org/[id]/overview/index.tsx:523]: Started a Slack conversation about this!

File: backend/src/services/org/org-service.ts
Comment (akhilmhdh) [reply backend/src/services/org/org-service.ts]: I am not talking about the login based email verification. Generally speaking if we run some migration and there existed an email validation checks its better the email above to valid ones rather than invalid emails

File: backend/src/ee/services/secret-approval-request/secret-approval-request-secret-dal.ts
Comment (akhilmhdh) [reply backend/src/ee/services/secret-approval-request/secret-approval-request-secret-dal.ts:26]: 100 is still in my opinion could be a bit trouble someone. If it was insert i mean 1000 wouldn't have matter. Where as 100 its individual query. N query this is not one. Where as the other one is 1 single query

File: backend/src/services/project-membership/project-membership-service.ts
Comment (akhilmhdh) [reply backend/src/services/project-membership/project-membership-service.ts]: `The function was like this before my changes as well`. Yes it's same as before but there is a one major change.

`for (const email of emails)`. This one is a new addition. This changed the whole process. The above is used by our frontend to add users to workspace on creation. If you don't need this change then kindly roll back the changes made in this file. Else kindly make the above changes, if the for loop exist

File: backend/src/server/routes/v1/project-router.ts
Comment (maidul98) [inline backend/src/server/routes/v1/project-router.ts:279]: Are you refering to invitee or inviter? Also is this relying on the frontend to always add the first user as the one who invites everyone?

File: backend/src/server/routes/v2/project-membership-router.ts
Comment (maidul98) [inline backend/src/server/routes/v2/project-membership-router.ts:12]: can you please add .describe on the zod parameters to briefly explain what the fields are? this will be helpful for api doc generation

File: backend/src/server/routes/v2/project-router.ts
Comment (maidul98) [inline backend/src/server/routes/v2/project-router.ts:29]: Looks like a breaking change?

File: backend/src/server/routes/v2/project-router.ts
Comment (maidul98) [inline backend/src/server/routes/v2/project-router.ts:127]: what about optional project slug?

Comment (maidul98) [reply backend/src/server/routes/v2/project-router.ts:127]: Also this API looks very very similar to the other project create. I can't seem to tell the difference other than naming parameters

File: backend/src/services/auth/auth-login-service.ts
Comment (maidul98) [inline backend/src/services/auth/auth-login-service.ts:278]: nit: this is true by default `isGhost: false`

File: backend/src/services/auth/auth-signup-service.ts
Comment (maidul98) [inline backend/src/services/auth/auth-signup-service.ts:53]: nit: isGhost: false is true by default

File: backend/src/services/org/org-dal.ts
Comment (maidul98) [inline backend/src/services/org/org-dal.ts:90]: filter out ghost user from this result too?

File: backend/src/services/org/org-service.ts
Comment (maidul98) [inline backend/src/services/org/org-service.ts:135]: what do you think about calling it `sudo-${alphaNumericNanoId(16)}-${orgId}@infisical.com`. It might sound a bit less scary and when we discuss this in the context of debugging with customers. Instead of saying `ghost` user we can refer to it as `sudo` user. This is your call though, just a suggestion

File: backend/src/services/project/project-queue.ts
Comment (maidul98) [inline backend/src/services/project/project-queue.ts:212]: I know there is a limit in terms of how many users someone can have on free tier, does this account for it?

File: backend/src/services/project/project-fns.ts
Comment (maidul98) [reply backend/src/services/project/project-fns.ts:8]: nit: `assignWorkspaceKeysToMembers` might better reflect what is going on in the function

File: backend/src/services/project-bot/project-bot-service.ts
Comment (maidul98) [inline backend/src/services/project-bot/project-bot-service.ts:123]: Nit: maybe we can just say Bot is already active

File: backend/src/services/secret/secret-service.ts
Comment (maidul98) [inline backend/src/services/secret/secret-service.ts:828]: nit: looks like this check is being repeated a lot, maybe you can put it in a function

Comment (maidul98) [reply backend/src/services/secret/secret-service.ts:828]: also hmm i tried to edit secrets while migration was happening and i didn't get any errors

File: backend/src/services/smtp/templates/workspaceInvitation.handlebars
Comment (maidul98) [inline backend/src/services/smtp/templates/workspaceInvitation.handlebars:12]: rephrase e2ee

**Review by maidul98 — COMMENTED** [review]
Looks real good! Just left some minor nits. Should be good to go after that

File: backend/src/ee/services/saml-config/saml-config-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/saml-config/saml-config-service.ts:342]: Not setting the field will lead to type errors. I suspect this is because I set the field to `notNullable`. Could this be an issue with the type system?

cc @akhilmhdh

File: backend/src/lib/crypto/srp.ts
Comment (DanielHougaard) [reply backend/src/lib/crypto/srp.ts:36]: Whoops! This is a leftover, good catch

File: backend/src/services/org/org-service.ts
Comment (DanielHougaard) [reply backend/src/services/org/org-service.ts:135]: I agree!

File: backend/src/services/project-membership/project-membership-service.ts
Comment (DanielHougaard) [reply backend/src/services/project-membership/project-membership-service.ts:403]:
```
const member = await userDAL.findUserByProjectMembershipId(membershipId);

    if (member?.isGhost) {
      throw new BadRequestError({
        message: "Unauthorized member delete",
        name: "Delete project membership"
      });
    }
```

Line 386-393

**Review by maidul98 — APPROVED** [review]

**Pull request merged**
